nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecord.java

```diff
@@ -0,0 +1,254 @@
+        .build();
+
+    static final PropertyDescriptor INDEX_RECORD_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-record-index-record-path")
+        .displayName("Index Record Path")
+        .description("A record path expression to retrieve the ID field for use with ElasticSearch. If left blank " +
+                "the index will be determined using the main index property.")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .build();
+
+    static final PropertyDescriptor TYPE_RECORD_PATH = new PropertyDescriptor.Builder()
+        .name("put-es-record-type-record-path")
+        .displayName("Type Record Path")
+        .description("A record path expression to retrieve the type field for use with ElasticSearch. If left blank " +
+                "the type will be determined using the main type property.")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .build();
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return new HashSet<Relationship>(){{
+            add(REL_SUCCESS);
+            add(REL_FAILURE);
+            add(REL_RETRY);
+        }};
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        List<PropertyDescriptor> _temp = new ArrayList<>();
+        _temp.add(INDEX);
+        _temp.add(TYPE);
+        _temp.add(CLIENT_SERVICE);
+        _temp.add(RECORD_READER);
+        _temp.add(OPERATION_RECORD_PATH);
+        _temp.add(ID_RECORD_PATH);
+        _temp.add(INDEX_RECORD_PATH);
+        _temp.add(TYPE_RECORD_PATH);
+
+        return Collections.unmodifiableList(_temp);
+    }
+
+    private RecordReaderFactory readerFactory;
+    private RecordPathCache recordPathCache;
+    private ElasticSearchClientService clientService;
+
+    @OnScheduled
+    public void onScheduled(ProcessContext context) {
+        this.readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        this.clientService = context.getProperty(CLIENT_SERVICE).asControllerService(ElasticSearchClientService.class);
+        this.recordPathCache = new RecordPathCache(16);
+    }
+
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if (input == null) {
+            return;
+        }
+
+        try {
+            List<IndexOperationRequest> bundle = buildOperations(input, context, session);
+            clientService.bulk(bundle);
+
+            session.transfer(input, REL_SUCCESS);
+        } catch (ElasticSearchError ese) {
+            String msg = String.format("Encountered a server-side problem with ElasticSearch. %s",
```

Comment:
Another style thing (not required, just FYI), the logging methods like `error` allow you to put empty braces in the string `{}` and follow it with an `Object[]` to pass in params like you are doing with `String.format`.

Reply:
I think I'll keep that one as-is because it's cleanly structured.